### PR TITLE
improve airbyte-cdk

### PR DESCRIPTION
**Pull request by stephane-airbyte**
various small changes related to tests and mssql:
the jdbc string for MSSQL was wrong. 
The LoggingInterceptor was printing faulty log lines
The SSHBastionContainer should be a container like any other, relying upon the ContainerFactory.
As a result, the ContainerFactory should be able to return any type of container, not just JdbcDatabaseContainer.
In addition, we shouldn't be passing strings to the ContainerFactory as container modifiers, but a named lambda. No more introspection needed! (the old string-based is kept for backward compatibility)

The CdcSourceTest doesn't check DB invariants before starting a source. We need a test function that can be overriden by the implementations of CdcSourceTest

### CODE REVIEW

**Comment by stephane-airbyte** [comment]
* **#35720** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/35720?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @stephane-airbyte and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/integrations/base/ssh/SshBastionContainer.java [hunk 1]
```diff
@@ -10,29 +10,50 @@
...
+          .withFileFromClasspath("Dockerfile", "bastion/Dockerfile"))
+              .withExposedPorts(22);
+      return container;
+    }
+
+    public GenericContainer exclusive(final Network network) {
+      Consumer<GenericContainer<?>> imageModifier = c -> {
+        c.withNetwork(network);
+      };
+      var container = super.exclusive("bastion-test", new NamedContainerModifierImpl<>("withNetwork", imageModifier));
```
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/integrations/base/ssh/SshBastionContainer.java:43]: create an actual method named withNetwork

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java [hunk 2]
```diff
@@ -85,9 +89,22 @@ private static final MdcScope.Builder getTestContainerLogMdcBuilder(DockerImageN
   /**
    * Returns a shared instance of the testcontainer.
    */
-  @SuppressWarnings("unchecked")
+  @Deprecated
```
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:92]: add a deprecation notice? e.g. "use foobarbaz() instead"

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java [hunk 3]
```diff
@@ -102,39 +119,79 @@ public final C shared(String imageName, String... methods) {
    * Returns an exclusive instance of the testcontainer.
    */
   @SuppressWarnings("unchecked")
+  @Deprecated
```
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:122]: same comment re. deprecation notice

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java [hunk 4]
```diff
@@ -102,39 +119,79 @@ public final C shared(String imageName, String... methods) {
...
   }
 
-  private GenericContainer<?> createAndStartContainer(DockerImageName imageName, List<String> methodNames) {
-    LOGGER.info("Creating new shared container based on {} with {}.", imageName, methodNames);
-    try {
-      GenericContainer<?> container = createNewContainer(imageName);
-      final var methods = new ArrayList<Method>();
-      for (String methodName : methodNames) {
-        methods.add(getClass().getMethod(methodName, container.getClass()));
+  public record NamedContainerModifierImpl<C extends GenericContainer<?>> (String name, Consumer<C> method) implements NamedContainerModifier<C> {
```
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:148]: Do we really want an interface that's only implemented once?

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java [hunk 5]
```diff
@@ -102,39 +119,79 @@ public final C shared(String imageName, String... methods) {
...
+
+    public String name() {
```
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:150]: Can you configure your editor to add `@Override` annotations please?

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java [hunk 6]
```diff
@@ -102,39 +119,79 @@ public final C shared(String imageName, String... methods) {
...
+    return resolvedMethod;
+  }
 
-        public void accept(OutputFrame frame) {
-          if (frame.getUtf8StringWithoutLineEnding().trim().length() > 0) {
-            super.accept(frame);
-          }
-        }
+  private C createAndStartContainer(DockerImageName imageName, List<? extends NamedContainerModifier<C>> namedContainerModifiers) {
+    LOGGER.info("Creating new shared container based on {} with {}.", imageName, Lists.transform(namedContainerModifiers, c -> c.name()));
```
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:175]: Fix the log message, it's no longer necessarily shared now.

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java (hunk 4 above)
Comment (stephane-airbyte) [reply airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:148]: it's not. it'll be implemented by the enums. See next PR in the stack

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java [hunk 7]
```diff
@@ -102,39 +119,79 @@ public final C shared(String imageName, String... methods) {
...
 
-      };
-      getTestContainerLogMdcBuilder(imageName, methodNames).produceMappings(logConsumer::withMdc);
-      container.withLogConsumer(logConsumer);
-      for (Method method : methods) {
-        LOGGER.info("Calling {} in {} on new shared container based on {}.",
-            method.getName(), getClass().getName(), imageName);
-        method.invoke(this, container);
+      public void accept(OutputFrame frame) {
+        if (frame.getUtf8StringWithoutLineEnding().trim().length() > 0) {
```
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:180]: Do we want to make some noise somehow in cases where this condition isn't met?

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/integrations/debezium/CdcSourceTest.java [hunk 8]
```diff
@@ -823,4 +834,7 @@ protected AirbyteCatalog expectedCatalogForDiscover() {
     return expectedCatalog;
   }
 
+  protected void waitForCdcRecords(String schemaName, String tableName, int recordCount)
+      throws Exception {}
```
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/integrations/debezium/CdcSourceTest.java:838]: If you want to be really pedantic about stacked changes shouldn't you be adding some kind of //TODO or //XXX comment here?

**Review by postamar — CHANGES_REQUESTED** [review]
Just a few little things

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java (hunk 5 above)
Comment (stephane-airbyte) [reply airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java:150]: It's already enabled. I don't think it applies on interface overrides

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/integrations/debezium/CdcSourceTest.java (hunk 8 above)
Comment (stephane-airbyte) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/integrations/debezium/CdcSourceTest.java:838]: that's not a todo. It's just a hook for other tests to insert some wait in here.

**Review by postamar — APPROVED** [review]

**Comment by stephane-airbyte** [comment]
### Merge activity

* **Feb 29, 4:30 PM EST**: @stephane-airbyte merged this pull request with [Graphite](https://app.graphite.dev/github/pr/airbytehq/airbyte/35720).

**Pull request merged**